### PR TITLE
Path table extraction

### PR DESCRIPTION
**Pull request by markisha64**
Some games have obfuscated file systems such that while the path table contains most info about directories,
those directories won't be accessible from the root directory.

I've added the pathTable option that instead of going recursively through the file system, goes recursively through the path table and reads files from folder offset.

This causes some issues during repacking, as since the file system was obfuscated, game probably reads files by going to sector offset. I've fixed this by supplying the exact offset that the file was read to in the xml file, and writing that file to the written offset (if it exists, otherwise calculated).

This mr is potentially very good for modding, as it will allow more games to be unpacked, modified and repacked. (currently only tested on dmw3)

My c++ isn't very good so if you see anything stinky, please do request changes.

### CODE REVIEW

File: src/dumpsxiso/main.cpp [hunk 1]
```diff
@@ -775,9 +872,15 @@ void ParseISO(cd::IsoReader& reader) {
     printf("\n");
 
     cd::IsoPathTable pathTable;
+		std::vector<cd::IsoPathTable::Entry> sorted;
 
     size_t numEntries = pathTable.ReadPathTable(&reader, descriptor.pathTable1Offs);
 
+		sorted = pathTable.pathTableList;
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:879]: Nitpick, but this sorted vector should be defined and assigned in one line, i.e.
```
std::vector<cd::IsoPathTable::Entry> sorted(pathTable.pathTableList);
```

File: src/dumpsxiso/main.cpp [hunk 2]
```diff
@@ -425,6 +427,79 @@ std::unique_ptr<cd::IsoDirEntries> ParseSubdirectory(cd::IsoReader& reader, List
...
+    } 
+
+    for (auto& e : dirEntries->dirEntryList.GetView()) {
+    		auto& entry = e.get();
+										
+    		entry.virtualPath = path;
+
+        if (entry.entry.flags & 0x2) {
+						int index = -1;
+						std::string s = "";
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:483]:
```suggestion
						std::string s;
```

**Review by CookiePLMonster — CHANGES_REQUESTED** [review]
The general idea seems solid, especially since it's hidden behind a flag as most people are unlikely to need it - but for some reason formatting is all over the place making it hard to read and review. Would be good to get this fixed before merging.

File: src/dumpsxiso/cdreader.cpp [hunk 3]
```diff
@@ -321,6 +321,36 @@ void cd::IsoDirEntries::ReadDirEntries(cd::IsoReader* reader, int lba, int secto
 		});
 }
 
+void cd::IsoDirEntries::ReadDirEntriesSkip(cd::IsoReader* reader, int lba, int sectors)
+{
+	size_t numEntries = 0; // Used to skip the first two entries, . and ..
```
[the remark this answers is not in the document]
Comment (CookiePLMonster) [reply src/dumpsxiso/cdreader.cpp:326]: Update - I just noticed it's the same in `cd::IsoDirEntries::ReadDirEntries` and you just copied it, my apologies - in this case you can leave it as-is and I'll check it at my own discretion, since I feel like I may have made a mistake there (depending on what the spec says).

[the remark this answers is not in the document]
Comment (CookiePLMonster) [reply src/dumpsxiso/cdreader.cpp:326]: Update2 - looks like the only difference between `ReadDirEntries` and `ReadDirEntriesSkip` is then `!(entry.value().entry.flags & 0x2))`? You could just introduce a new boolean parameter to `ReadDirEntries` then instead of duplicating the entire function.

File: src/mkpsxiso/iso.cpp [hunk 4]
```diff
@@ -354,18 +358,22 @@ int iso::DirTreeClass::CalculateTreeLBA(int lba)
...
+				lba += (entry.flba)
+					? entry.flba - lba + GetSizeInSectors(entry.length, 2048)
+					: GetSizeInSectors(entry.length, 2048);					
 			}
 			else if ( entry.type == EntryType::EntryXA )
 			{
-				lba += GetSizeInSectors(entry.length, 2336);
+				lba += (entry.flba)
+					? entry.flba - lba + GetSizeInSectors(entry.length, 2336)
+					: GetSizeInSectors(entry.length, 2336);
```
[the remark this answers is not in the document]
Comment (markisha64) [reply src/mkpsxiso/iso.cpp:376]: yes, esentially this is lbe = entry.flba + size if it exists, I wasn't sure what would happen if lba wasn't properly tracked (even here I'm not sure) so instead of just adding size to the end I'm also moving it to the last flba, tho in retrospect, this might still be a problem if the last files isn't the one with the highest offset

File: src/dumpsxiso/cdreader.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (markisha64) [reply src/dumpsxiso/cdreader.cpp:326]: Yeah, in retrospect, I should've just made it a boolean parameter, I'll do that

**Comment by markisha64** [comment]
> but for some reason formatting is all over the place making it hard to read and review
this is kind of a project wide problem, the tabs and indentations are all over the place and when I started it was really hard to follow, something should be eventually done about this, but it's out of scope for the pr

File: src/dumpsxiso/main.cpp [hunk 5]
```diff
@@ -425,6 +427,79 @@ std::unique_ptr<cd::IsoDirEntries> ParseSubdirectory(cd::IsoReader& reader, List
 	return dirEntries;
 }
 
+std::unique_ptr<cd::IsoDirEntries> ParsePathTable(cd::IsoReader& reader, ListView<cd::IsoDirEntries::Entry> view, std::vector<cd::IsoPathTable::Entry>& pathTableList, std::vector<cd::IsoPathTable::Entry>& sorted, int index,
+   const fs::path& path) {
+    auto dirEntries = std::make_unique<cd::IsoDirEntries>(std::move(view));
+
+		int sec = pathTableList[index].entry.extLength 
+				? pathTableList[index].entry.extLength 
+				: 1;
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:436]: This is probably better done as `std::min(1, pathTableList[index].entry.extLength);`

File: src/dumpsxiso/main.cpp [hunk 6]
```diff
@@ -595,6 +683,7 @@ tinyxml2::XMLElement* WriteXMLEntry(const cd::IsoDirEntries::Entry& entry, tinyx
 {
 	tinyxml2::XMLElement* newelement;
 
+
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:686]: Stray newline.

File: src/dumpsxiso/cdreader.h [hunk 7]
```diff
@@ -100,9 +100,10 @@ namespace cd {
         ListView<Entry> dirEntryList;
 
         IsoDirEntries(ListView<Entry> view);
-        void ReadDirEntries(cd::IsoReader* reader, int lba, int sectors);
+        void ReadDirEntries(cd::IsoReader* reader, int lba, int sectors, bool skipFolders);
         void ReadRootDir(cd::IsoReader* reader, int lba);
 
+
```
Comment (CookiePLMonster) [inline src/dumpsxiso/cdreader.h:106]: Stray newline.

File: src/dumpsxiso/main.cpp [hunk 8]
```diff
@@ -778,6 +875,11 @@ void ParseISO(cd::IsoReader& reader) {
 
     size_t numEntries = pathTable.ReadPathTable(&reader, descriptor.pathTable1Offs);
 
+		std::vector<cd::IsoPathTable::Entry> sorted(pathTable.pathTableList);
+		std::sort(sorted.begin(), sorted.end(), [](auto& a, auto& b) {
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:879]:
```suggestion
		std::sort(sorted.begin(), sorted.end(), [](const auto& a, const auto& b) {
```

AFAIK non-const predicate parameters might not compile on all compilers, better safe than sorry.

File: src/dumpsxiso/main.cpp [hunk 9]
```diff
@@ -964,7 +1068,8 @@ int Main(int argc, char *argv[])
 		"  -s <path>  - Outputs an MKPSXISO compatible XML script for later rebuilding.\n"
 		"  -S|--sort-by-dir - Outputs a \"pretty\" XML script where entries are grouped in directories, instead of strictly following their original order on the disc.\n"
 		"  -e|--encode <codec> - Codec to encode CDDA/DA audio. wave is default. Supported codecs: " SUPPORTED_CODEC_TEXT "\n"
-		"  -h|--help  - Show this help text\n";
+		"  -h|--help  - Show this help text\n"
+		"  -pt|--pathTable - instead of going throught the file system, go to every known directory in order; helps with deobfuscating\n";
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:1072]:
```suggestion
		"  -pt|--path-table - instead of going through the file system, go to every known directory in order; helps with deobfuscating\n";
```

`sort-by-dir` sets a precedent to use `this-style-of-naming` over `thisStyle`. Let's be consistent.

File: src/dumpsxiso/main.cpp [hunk 10]
```diff
@@ -982,6 +1087,11 @@ int Main(int argc, char *argv[])
 		// Is it a switch?
 		if ((*args)[0] == '-')
 		{
+			if (ParseArgument(args, "pt", "pathTable"))
```
Comment (CookiePLMonster) [inline src/dumpsxiso/main.cpp:1090]:
```suggestion
			if (ParseArgument(args, "pt", "path-table"))
```

See above.

**Review by CookiePLMonster — CHANGES_REQUESTED** [review]
Generally looks fine now, a few final comments from me and after that I'll approve. I poked @spicyjpeg to also have a look.

File: src/dumpsxiso/main.cpp (hunk 5 above)
Comment (markisha64) [reply src/dumpsxiso/main.cpp:436]: Pretty sure that wouldn't work, because 0 wouldn't read the directory (and I've found that in such obfuscated fs-es they leave it 0 sometimes)
If you want is as a one line `std::max(1, pathTableList[index].entry.extLength)` would work

Comment (CookiePLMonster) [reply src/dumpsxiso/main.cpp:436]: Of course, I got confused - I meant `max` indeed.

File: src/mkpsxiso/iso.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (markisha64) [reply src/mkpsxiso/iso.cpp]: Added a safer way to track final lba with offsets

**Review by CookiePLMonster — APPROVED** [review]
Looks good now. I'll run it against my test suite and merge if nothing regressed.

**Comment by CookiePLMonster** [comment]
All tests passed as before - I've not checked dumpsxiso but it should be fine too.
